File: src/connectors/local_bridge.zig
//! Local inference bridge — a Zig HTTP client that dispatches completions to a
//! user-run local OpenAI-compatible server (llama-server, ollama, mlx-server,
//! lm-server, etc.). ABI does not embed or bundle any external inference engine;
//! the user starts the server separately and ABI talks to it over loopback HTTP.
//!
//! The bridge endpoint is configured via environment variables:
//!   `ABI_LLAMA_CPP_ENDPOINT` (default `http://127.0.0.1:8080`)
//!   `ABI_MLX_ENDPOINT`       (default `http://127.0.0.1:8081`)
//!
//! Model id prefixes routed to this bridge: `llama-cpp/`, `llama/`, `ollama/`,
//! `ollama-`, `lmstudio/`, `vllm/`, `mlx/`, `mlx-`.

const std = @import("std");
const connector = @import("connector.zig");
const http = @import("http.zig");
const json = @import("json.zig");

const ConnectorError = connector.ConnectorError;
const ConnectorConfig = connector.ConnectorConfig;
const Response = connector.Response;

/// Default endpoints for known local inference servers.
pub const LLAMA_CPP_DEFAULT_ENDPOINT = "http://127.0.0.1:8080";
pub const MLX_DEFAULT_ENDPOINT = "http://127.0.0.1:8081";

/// Re-export HTTP stream types for callers
pub const StreamCallback = http.StreamCallback;
pub const StreamChunk = http.StreamChunk;

/// Determine whether a model id should dispatch to a local inference bridge
/// rather than the in-process persona router. Matches the same prefixes that
/// `models.providerOf` classifies as `.local`.
pub fn isLocalBridgeModel(model: []const u8) bool {
    const prefixes = [_][]const u8{
        "llama-cpp/", "llama/", "ollama/", "ollama-",
        "lmstudio/",  "vllm/",  "mlx/",    "mlx-",
    };
    for (prefixes) |prefix| {
        if (std.mem.startsWith(u8, model, prefix)) return true;
    }
    return false;
}

/// Resolve the bridge endpoint for a model id. `mlx/` and `mlx-` prefixes use
/// the MLX endpoint; all other local-bridge prefixes use the llama-cpp endpoint.
/// Callers can pass an override endpoint (e.g. from env) via `override`.
pub fn endpointFor(model: []const u8, override: ?[]const u8) []const u8 {
    if (override) |ep| return ep;
    if (std.mem.startsWith(u8, model, "mlx/") or std.mem.startsWith(u8, model, "mlx-")) {
        return MLX_DEFAULT_ENDPOINT;
    }
    return LLAMA_CPP_DEFAULT_ENDPOINT;
}

/// Check whether the local inference server is reachable. Returns `false`
/// (not an error) when the server is down — callers should fall back to the
/// in-process persona router with a warning.
pub fn healthCheck(io: std.Io, allocator: std.mem.Allocator, endpoint: []const u8) bool {
    const config = ConnectorConfig{
        .api_key = "",
        .base_url = endpoint,
        .timeout_ms = 3000,
        .transport = .live,
    };
    var response = http.httpGetJson(io, allocator, config, "/health") catch return false;
    defer response.deinit(allocator);
    return response.status == 200;
}

/// Send a completion request to the local inference server. The request body
/// is OpenAI-compatible (`/v1/chat/completions`). Returns the response body
/// (owned by the caller). The caller is responsible for parsing the
/// `choices[0].message.content` field.
pub fn completeLive(
    io: std.Io,
    allocator: std.mem.Allocator,
    model: []const u8,
    input: []const u8,
) !Response {
    const endpoint = endpointFor(model, null);
    const config = ConnectorConfig{
        .api_key = "",
        .base_url = endpoint,
        .timeout_ms = 30000,
        .transport = .live,
    };

    // Build OpenAI-compatible messages array
    var messages = std.ArrayListUnmanaged(u8).empty;
    defer messages.deinit(allocator);
    try messages.appendSlice(allocator, "[{\"role\":\"user\",\"content\":");
    try json.appendJsonString(&messages, allocator, input);
    try messages.appendSlice(allocator, "}]");

    const body = try json.buildOpenAiBody(allocator, model, messages.items, false);
    defer allocator.free(body);

    return try http.httpPostJson(io, allocator, config, "/v1/chat/completions", body, &.{});
}

/// Send a streaming completion request to the local inference server via SSE.
/// The server must support OpenAI-compatible `/v1/chat/completions` with
/// `stream=true`. The callback is invoked for each token delta.
pub fn completeLiveStreaming(
    io: std.Io,
    allocator: std.mem.Allocator,
    model: []const u8,
    input: []const u8,
    on_chunk: http.StreamCallback,
    callback_ctx: *anyopaque,
) ![]const u8 {
    const endpoint = endpointFor(model, null);
    const config = ConnectorConfig{
        .api_key = "",
        .base_url = endpoint,
        .timeout_ms = 60000,
        .transport = .live,
    };

    // Build OpenAI-compatible messages array with stream=true
    var messages = std.ArrayListUnmanaged(u8).empty;
    defer messages.deinit(allocator);
    try messages.appendSlice(allocator, "[{\"role\":\"user\",\"content\":");
    try json.appendJsonString(&messages, allocator, input);
    try messages.appendSlice(allocator, "}]");

    const body = try json.buildOpenAiBody(allocator, model, messages.items, true);
    defer allocator.free(body);

    return try http.httpPostJsonStreamingIncremental(io, allocator, config, "/v1/chat/completions", body, on_chunk, callback_ctx, &.{});
}

/// Extract the completion text from an OpenAI-compatible JSON response.
/// Returns an owned slice. Returns `error.InvalidResponse` if the response
/// is not valid OpenAI JSON or the `choices` array is empty.
pub fn extractCompletion(allocator: std.mem.Allocator, body: []const u8) ![]u8 {
    const parsed = std.json.parseFromSlice(std.json.Value, allocator, body, .{}) catch return ConnectorError.InvalidResponse;
    defer parsed.deinit();
    const root = parsed.value.object;
    const choices = root.get("choices") orelse return ConnectorError.InvalidResponse;
    if (choices.array.items.len == 0) return ConnectorError.InvalidResponse;
    const choice = choices.array.items[0].object;
    const message = choice.get("message") orelse return ConnectorError.InvalidResponse;
    const content = message.object.get("content") orelse return ConnectorError.InvalidResponse;
    return try allocator.dupe(u8, content.string);
}

test "isLocalBridgeModel matches known prefixes" {
    try std.testing.expect(isLocalBridgeModel("llama-cpp/mistral"));
    try std.testing.expect(isLocalBridgeModel("llama/llama3"));
    try std.testing.expect(isLocalBridgeModel("ollama/qwen2"));
    try std.testing.expect(isLocalBridgeModel("ollama-qwen2"));
    try std.testing.expect(isLocalBridgeModel("lmstudio/phi3"));
    try std.testing.expect(isLocalBridgeModel("vllm/mixtral"));
    try std.testing.expect(isLocalBridgeModel("mlx/gemma"));
    try std.testing.expect(isLocalBridgeModel("mlx-gemma"));
    try std.testing.expect(!isLocalBridgeModel("claude-fable-5"));
    try std.testing.expect(!isLocalBridgeModel("abi-local"));
    try std.testing.expect(!isLocalBridgeModel("gpt-4"));
}

test "fable-5 alias is not a local bridge model" {
    // The catalog alias `fable-5` (-> `claude-fable-5`) must never match the
    // local-bridge prefix table — it routes to the anthropic provider, not here.
    try std.testing.expect(!isLocalBridgeModel("fable-5"));
    try std.testing.expect(!isLocalBridgeModel("fable5"));
}

test "endpointFor routes mlx to MLX endpoint" {
    try std.testing.expectEqualStrings(LLAMA_CPP_DEFAULT_ENDPOINT, endpointFor("llama-cpp/mistral", null));
    try std.testing.expectEqualStrings(MLX_DEFAULT_ENDPOINT, endpointFor("mlx/gemma", null));
    try std.testing.expectEqualStrings(MLX_DEFAULT_ENDPOINT, endpointFor("mlx-gemma", null));
    try std.testing.expectEqualStrings(LLAMA_CPP_DEFAULT_ENDPOINT, endpointFor("ollama/qwen2", null));
    // Override takes precedence
    try std.testing.expectEqualStrings("http://localhost:9999", endpointFor("mlx/gemma", "http://localhost:9999"));
}

test "extractCompletion parses OpenAI response" {
    const allocator = std.testing.allocator;
    const body =
        \\{"id":"chatcmpl-1","object":"chat.completion","choices":[{"index":0,"message":{"role":"assistant","content":"Hello world"},"finish_reason":"stop"}]}
    ;
    const result = try extractCompletion(allocator, body);
    defer allocator.free(result);
    try std.testing.expectEqualStrings("Hello world", result);
}

test "extractCompletion rejects empty choices" {
    const allocator = std.testing.allocator;
    const body = "{\"choices\":[]}";
    try std.testing.expectError(ConnectorError.InvalidResponse, extractCompletion(allocator, body));
}

test "extractCompletion rejects invalid JSON" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(ConnectorError.InvalidResponse, extractCompletion(allocator, "not json"));
}

test {
    std.testing.refAllDecls(@This());
}

File: crates/abi-connectors/src/local_bridge.rs
//! Local OpenAI-compatible inference bridge.
//!
//! Ported from `src/connectors/local_bridge.zig`. Dispatches completions to a
//! user-run local server (llama-server, ollama, mlx-server, …) over loopback
//! HTTP. ABI does **not** embed an inference engine.
//!
//! Endpoints (env overrides):
//! - [`LLAMA_CPP_DEFAULT_ENDPOINT`] / `ABI_LLAMA_CPP_ENDPOINT`
//! - [`MLX_DEFAULT_ENDPOINT`] / `ABI_MLX_ENDPOINT`

use crate::connector::{ConnectorConfig, ConnectorError, Response, Result, TransportMode};
use crate::payload::build_openai_body;
use crate::transport::{Method, Transport, execute};
use abi_foundation::env::{self, LLAMA_CPP_ENDPOINT, MLX_ENDPOINT};
use serde_json::Value;

/// Default llama.cpp / ollama-style endpoint.
pub const LLAMA_CPP_DEFAULT_ENDPOINT: &str = "http://127.0.0.1:8080";
/// Default MLX server endpoint.
pub const MLX_DEFAULT_ENDPOINT: &str = "http://127.0.0.1:8081";

const LOCAL_PREFIXES: &[&str] = &[
    "llama-cpp/",
    "llama/",
    "ollama/",
    "ollama-",
    "lmstudio/",
    "vllm/",
    "mlx/",
    "mlx-",
];

/// Whether `model` should dispatch to the local bridge (not the persona router).
#[must_use]
pub fn is_local_bridge_model(model: &str) -> bool {
    LOCAL_PREFIXES.iter().any(|p| model.starts_with(p))
}

/// Resolve the bridge base URL for `model`.
///
/// `override_endpoint` wins when set. Otherwise `mlx/` / `mlx-` use the MLX
/// endpoint (env or default); other local models use the llama.cpp endpoint.
#[must_use]
pub fn endpoint_for(model: &str, override_endpoint: Option<&str>) -> String {
    if let Some(ep) = override_endpoint.filter(|s| !s.is_empty()) {
        return ep.to_string();
    }
    let is_mlx = model.starts_with("mlx/") || model.starts_with("mlx-");
    if is_mlx {
        env::get(MLX_ENDPOINT).unwrap_or_else(|| MLX_DEFAULT_ENDPOINT.to_string())
    } else {
        env::get(LLAMA_CPP_ENDPOINT).unwrap_or_else(|| LLAMA_CPP_DEFAULT_ENDPOINT.to_string())
    }
}

fn live_config(endpoint: &str, timeout_ms: u32) -> ConnectorConfig {
    ConnectorConfig {
        api_key: "local-bridge".into(), // some servers require a non-empty key header
        base_url: endpoint.trim_end_matches('/').to_string(),
        timeout_ms,
        transport: TransportMode::Live,
    }
}

/// `GET /health` on `endpoint`. Returns `false` when unreachable (not an error).
pub fn health_check(transport: &dyn Transport, endpoint: &str) -> bool {
    let config = live_config(endpoint, 3_000);
    let Ok(request) =
        crate::transport::build_request(&config, Method::Get, "/health", None, None, Vec::new())
    else {
        return false;
    };
    match execute(transport, &request) {
        Ok(response) => response.status == 200,
        Err(_) => false,
    }
}

/// Non-streaming OpenAI-compatible completion against the local bridge.
pub fn complete_live(
    transport: &dyn Transport,
    model: &str,
    input: &str,
    endpoint_override: Option<&str>,
) -> Result<Response> {
    let endpoint = endpoint_for(model, endpoint_override);
    let config = live_config(&endpoint, 30_000);
    let messages = build_user_messages(input);
    let body = build_openai_body(model, &messages, false)?;
    let request = crate::transport::build_request(
        &config,
        Method::Post,
        "/v1/chat/completions",
        Some("application/json"),
        Some(body),
        Vec::new(),
    )?;
    execute(transport, &request)
}

/// Streaming OpenAI-compatible completion; returns the raw SSE body.
pub fn complete_live_streaming(
    transport: &dyn Transport,
    model: &str,
    input: &str,
    endpoint_override: Option<&str>,
) -> Result<Response> {
    let endpoint = endpoint_for(model, endpoint_override);
    let config = live_config(&endpoint, 60_000);
    let messages = build_user_messages(input);
    let body = build_openai_body(model, &messages, true)?;
    let request = crate::transport::build_request(
        &config,
        Method::Post,
        "/v1/chat/completions",
        Some("application/json"),
        Some(body),
        Vec::new(),
    )?;
    execute(transport, &request)
}

fn build_user_messages(input: &str) -> String {
    let mut out = String::from(r#"[{"role":"user","content":"#);
    out.push_str(&abi_foundation::json::escape_json_string(input));
    out.push_str("}]");
    out
}

/// Extract `choices[0].message.content` from an `OpenAI` chat completion body.
pub fn extract_completion(body: &str) -> Result<String> {
    let parsed: Value = serde_json::from_str(body).map_err(|_| ConnectorError::InvalidResponse)?;
    let content = parsed
        .get("choices")
        .and_then(|c| c.as_array())
        .and_then(|a| a.first())
        .and_then(|c| c.get("message"))
        .and_then(|m| m.get("content"))
        .and_then(Value::as_str)
        .ok_or(ConnectorError::InvalidResponse)?;
    Ok(content.to_string())
}

/// Convenience: health-check via the default live HTTP transport.
#[cfg(feature = "live")]
#[must_use]
pub fn health_check_default(endpoint: &str) -> bool {
    health_check(&crate::transport::HttpTransport::new(), endpoint)
}

/// Convenience: complete via the default live HTTP transport.
#[cfg(feature = "live")]
pub fn complete_live_default(
    model: &str,
    input: &str,
    endpoint_override: Option<&str>,
) -> Result<Response> {
    complete_live(
        &crate::transport::HttpTransport::new(),
        model,
        input,
        endpoint_override,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::transport::RecordingTransport;

    #[test]
    fn is_local_bridge_model_matches_known_prefixes() {
        assert!(is_local_bridge_model("llama-cpp/mistral"));
        assert!(is_local_bridge_model("llama/llama3"));
        assert!(is_local_bridge_model("ollama/qwen2"));
        assert!(is_local_bridge_model("ollama-qwen2"));
        assert!(is_local_bridge_model("lmstudio/phi3"));
        assert!(is_local_bridge_model("vllm/mixtral"));
        assert!(is_local_bridge_model("mlx/gemma"));
        assert!(is_local_bridge_model("mlx-gemma"));
        assert!(!is_local_bridge_model("claude-fable-5"));
        assert!(!is_local_bridge_model("abi-local"));
        assert!(!is_local_bridge_model("gpt-4"));
        assert!(!is_local_bridge_model("fable-5"));
    }

    #[test]
    fn endpoint_for_routes_mlx_and_override() {
        assert_eq!(
            endpoint_for("llama-cpp/mistral", None),
            LLAMA_CPP_DEFAULT_ENDPOINT
        );
        assert_eq!(endpoint_for("mlx/gemma", None), MLX_DEFAULT_ENDPOINT);
        assert_eq!(endpoint_for("mlx-gemma", None), MLX_DEFAULT_ENDPOINT);
        assert_eq!(
            endpoint_for("ollama/qwen2", None),
            LLAMA_CPP_DEFAULT_ENDPOINT
        );
        assert_eq!(
            endpoint_for("mlx/gemma", Some("http://localhost:9999")),
            "http://localhost:9999"
        );
    }

    #[test]
    fn extract_completion_parses_openai_response() {
        let body = r#"{"id":"chatcmpl-1","choices":[{"index":0,"message":{"role":"assistant","content":"Hello world"},"finish_reason":"stop"}]}"#;
        assert_eq!(extract_completion(body).unwrap(), "Hello world");
        assert!(extract_completion(r#"{"choices":[]}"#).is_err());
        assert!(extract_completion("not json").is_err());
    }

    #[test]
    fn complete_live_builds_openai_request_against_override() {
        let transport = RecordingTransport::new().with_response(Ok(Response::ok(
            r#"{"choices":[{"message":{"content":"pong"}}]}"#.to_string(),
        )));
        let response = complete_live(&transport, "llama/phi3", "ping", Some("http://127.0.0.1:1"))
            .expect("synthetic transport");
        assert_eq!(response.status, 200);
        let requests = transport.requests();
        assert_eq!(requests.len(), 1);
        assert!(requests[0].url.contains("127.0.0.1:1"));
        assert!(requests[0].url.ends_with("/v1/chat/completions"));
        let body = requests[0].body.as_deref().unwrap_or("");
        assert!(body.contains("\"model\":\"llama/phi3\""));
        assert!(body.contains("ping"));
    }

    #[test]
    fn health_check_false_on_transport_error() {
        use crate::transport::UnavailableTransport;
        assert!(!health_check(&UnavailableTransport, "http://127.0.0.1:1"));
    }
}
